File: Languages and Automata I/lmu-cmsi3801/python/tests/exercises_test.py
import re
import math
import pytest
from exercises import (
    change, stretched, powers, say, find_first_then_lower,
    top_ten_scorers, turing_machine)


def test_change():
    assert change(0) == (0, 0, 0, 0)
    assert change(97) == (3, 2, 0, 2)
    assert change(8) == (0, 0, 1, 3)
    assert change(250) == (10, 0, 0, 0)
    assert change(144) == (5, 1, 1, 4)
    assert change(97) == (3, 2, 0, 2)
    assert change(100000000000) == (4000000000, 0, 0, 0)
    with pytest.raises(TypeError) as excinfo:
        # No fractional amounts allowed
        change(988.25)
    with pytest.raises(TypeError) as excinfo:
        # Do type check before negative check
        change(-988.25)
    with pytest.raises(ValueError) as excinfo:
        change(-50)
    assert str(excinfo.value) == 'amount cannot be negative'


def test_stretched():
    assert stretched('') == ''
    assert stretched('H e   l\t\tlo') == 'Heelllllllooooo'
    assert stretched('$#') == '$##'
    assert stretched('       ') == ''
    assert stretched('A = πr²') == 'A==πππrrrr²²²²²'
    assert stretched("  ") == ""
    assert stretched("  \t\n  \t") == ""
    assert stretched("  Hi  hi  ") == "Hiihhhiiii"
    assert stretched("😁😂😱") == "😁😂😂😱😱😱"
    assert stretched("hello world") == "heelllllllooooowwwwwwooooooorrrrrrrrllllllllldddddddddd"


def test_powers():
    p = powers(base=2, limit=10)
    assert next(p) == 1
    assert next(p) == 2
    assert next(p) == 4
    assert next(p) == 8
    with pytest.raises(StopIteration):
        next(p)
    assert list(powers(base=2, limit=-5)) == []
    assert list(powers(base=7, limit=0)) == []
    assert list(powers(limit=1, base=3)) == [1]
    assert list(powers(base=2, limit=63)) == [1, 2, 4, 8, 16, 32]
    assert list(powers(base=2, limit=64)) == [1, 2, 4, 8, 16, 32, 64]
    with pytest.raises(TypeError):
        # Test that the arguments MUST be keyword arguments
        powers(2, 100)


def test_say():
    assert say() == ''
    assert say('hi')() == 'hi'
    assert say('hi')('there')() == 'hi there'
    assert say('hello')('my')('name')('is')(
        'Colette')() == 'hello my name is Colette'


def test_find_first_then_lower_case():
    assert find_first_then_lower(lambda s: len(s) > 10, []) is None
    assert find_first_then_lower(lambda s: len(s) > 5, ["hello", "world"]) is None
    assert find_first_then_lower(lambda s: s.startswith("HELL"), ["HELLO", "WORLD"]) == "hello"
    assert find_first_then_lower(lambda s: "d!" in s, ["Hello", "World!!"]) == "world!!"


def test_top_ten_scorers():
    assert top_ten_scorers({}) == []
    assert top_ten_scorers({'T1': [['A', 3, 300]]}) == []
    assert top_ten_scorers({'T1': [['A', 30, 300]]}) == ['A|10.00|T1']
    input = {
        'ATL': [
            ['Betnijah Laney', 16, 263],
            ['Courtney Williams', 14, 193]],
        'CHI': [
            ['Kahleah Copper', 17, 267],
            ['Allie Quigley', 17, 260],
            ['Courtney Vandersloot', 17, 225]],
        'CONN': [
            ['DeWanna Bonner', 16, 285],
            ['Alyssa Thomas', 16, 241]],
        'DAL': [
            ['Arike Ogunbowale', 16, 352],
            ['Satou Sabally', 12, 153]],
        'IND': [
            ['Kelsey Mitchell', 16, 280],
            ['Tiffany Mitchell', 13, 172],
            ['Candice Dupree', 16, 202]],
        'LA': [
            ['Nneka Ogwumike', 14, 172],
            ['Chelsea Gray', 16, 224],
            ['Candace Parker', 16, 211]],
        'LV': [
            ['A’ja Wilson', 15, 304],
            ['Dearica Hamby', 15, 188],
            ['Angel McCoughtry', 15, 220]],
        'MIN': [
            ['Napheesa Collier', 16, 262],
            ['Crystal Dangerfield', 16, 254]],
        'NY': [
            ['Layshia Clarendon', 15, 188]],
        'PHX': [
            ['Diana Taurasi', 13, 236],
            ['Brittney Griner', 12, 212],
            ['Skylar Diggins-Smith', 16, 261],
            ['Bria Hartley', 13, 190]],
        'SEA': [
            ['Breanna Stewart', 16, 317],
            ['Jewell Loyd', 16, 223]],
        'WSH': [
            ['Emma Meesseman', 13, 158],
            ['Ariel Atkins', 15, 212],
            ['Myisha Hines-Allen', 15, 236]]}
    expected = [
        "Arike Ogunbowale|22.00|DAL",
        "A’ja Wilson|20.27|LV",
        "Breanna Stewart|19.81|SEA",
        "DeWanna Bonner|17.81|CONN",
        "Kelsey Mitchell|17.50|IND",
        "Betnijah Laney|16.44|ATL",
        "Napheesa Collier|16.38|MIN",
        "Skylar Diggins-Smith|16.31|PHX",
        "Crystal Dangerfield|15.88|MIN",
        "Myisha Hines-Allen|15.73|WSH"
    ]
    assert top_ten_scorers(input) == expected


def test_turing_machine():
    # TM description is one rule per line: cur_state, read, write, move, next_state
    times_four = """Scan,0,0,R,Scan
        Scan,1,1,R,Scan
        Scan,#,0,R,Adding
        Adding,#,0,R,Halt"""
    assert list(turing_machine(times_four, "1010")) == [
        ('', 'Scan', '1010'),
        ('1', 'Scan', '010'),
        ('10', 'Scan', '10'),
        ('101', 'Scan', '0'),
        ('1010', 'Scan', ''),
        ('10100', 'Adding', ''),
        ('101000', 'Halt', '')]
    assert list(turing_machine(times_four, "1")) == [
        ('', 'Scan', '1'),
        ('1', 'Scan', ''),
        ('10', 'Adding', ''),
        ('100', 'Halt', '')]
    even_test = """S,0,0,R,S
        S,1,1,R,S
        S,#,#,L,End
        End,0,0,R,Accept
        End,1,1,R,Reject
        End,#,#,R,Reject"""
    assert list(turing_machine(even_test, "101")) == [
        ('', 'S', '101'),
        ('1', 'S', '01'),
        ('10', 'S', '1'),
        ('101', 'S', ''),
        ('10', 'End', '1#'),
        ('101', 'Reject', '#')]
    incrementer = """S,0,0,R,S
        S,1,1,R,S
        S,#,#,L,B
        B,1,0,L,B
        B,0,1,L,H
        B,#,#,L,H"""
    assert list(turing_machine(incrementer, "101")) == [
        ('', 'S', '101'),
        ('1', 'S', '01'),
        ('10', 'S', '1'),
        ('101', 'S', ''),
        ('10', 'B', '1#'),
        ('1', 'B', '00#'),
        ('', 'H', '110#')]

File: Languages and Automata I/lmu-cmsi3801/java/ExercisesTest.java
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;

public class ExercisesTest {

    @Test
    public void testChange() {
        assertThrows(IllegalArgumentException.class, () -> Exercises.change(-20));
        assertEquals(Exercises.change(0), List.of(0, 0, 0, 0));
        assertEquals(Exercises.change(1), List.of(0, 0, 0, 1));
        assertEquals(Exercises.change(13), List.of(0, 1, 0, 3));
        assertEquals(Exercises.change(50), List.of(2, 0, 0, 0));
        assertEquals(Exercises.change(5), List.of(0, 0, 1, 0));
        assertEquals(Exercises.change(397), List.of(15, 2, 0, 2));
        assertEquals(Exercises.change(42), List.of(1, 1, 1, 2));
        assertEquals(Exercises.change(1000000017), List.of(40000000, 1, 1, 2));
    }

    @Test
    public void testStretched() {
        assertEquals(Exercises.stretched(""), "");
        assertEquals(Exercises.stretched("  "), "");
        assertEquals(Exercises.stretched("  \t\n  \t"), "");
        assertEquals(Exercises.stretched("  Hi  hi  "), "Hiihhhiiii");
        assertEquals(Exercises.stretched("😁😂😱"), "😁😂😂😱😱😱");
        assertEquals(Exercises.stretched("hello world"), "heelllllllooooowwwwwwooooooorrrrrrrrllllllllldddddddddd");
    }

    @Test
    public void testPowersStream() {
        assertArrayEquals(Exercises.powers(1).limit(5).toArray(),
            new int[] { 1, 1, 1, 1, 1 });
        assertArrayEquals(Exercises.powers(7).limit(10).toArray(),
            new int[] { 1, 7, 49, 343, 2401, 16807, 117649, 823543, 5764801, 40353607 });
        assertArrayEquals(Exercises.powers(-3).limit(5).toArray(),
            new int[] { 1, -3, 9, -27, 81 });
        assertArrayEquals(Exercises.powers(10).limit(4).toArray(),
            new int[] { 1, 10, 100, 1000 });
    }

    @Test
    public void testSay() {
        assertEquals(Exercises.say("A").ok(), "A");
        assertEquals(Exercises.say(), "");
        assertEquals(Exercises.say("A").and("B").ok(), "A B");
        assertEquals(Exercises.say("🐤🦇").and("$🦊👏🏽").and("!").ok(), "🐤🦇 $🦊👏🏽 !");
    }

    @Test
    public void testFindFirstAndLowerCase() {
        assertEquals(
            Exercises.findFirstThenLower(s -> s.length() > 10, List.of()),
            Optional.empty()
        );
        assertEquals(
            Exercises.findFirstThenLower(s -> s.length() > 5, List.of("hello", "world")),
            Optional.empty()
        );
        assertEquals(
            Exercises.findFirstThenLower(s -> s.startsWith("HELL"), List.of("HELLO", "WORLD")),
            Optional.of("hello")
        );
        assertEquals(
            Exercises.findFirstThenLower(s -> s.contains("d!"), List.of("Hello", "World!!")),
            Optional.of("world!!")
        );
    }

    @Test
    public void testTopTenScorers() {
        assertEquals(Exercises.topTenScorers(Map.of()), List.of());

        assertEquals(
            Exercises.topTenScorers(Map.of("T1", List.of("A,30,300"))),
            List.of("A|10.00|T1")
        );

        var stats = new HashMap<String, List<String>>();
        stats.put("ATL", List.of(
            "Betnijah Laney,16,263", "Courtney Williams,14,193"));
        stats.put("CHI", List.of(
            "Kahleah Copper,17,267", "Allie Quigley,17,260", "Courtney Vandersloot,17,225"));
        stats.put("CONN", List.of(
            "DeWanna Bonner,16,285", "Alyssa Thomas,16,241"));
        stats.put("DAL", List.of(
            "Arike Ogunbowale,16,352", "Satou Sabally,12,153"));
        stats.put("IND", List.of(
            "Kelsey Mitchell,16,280", "Tiffany Mitchell,13,172", "Candice Dupree,16,202"));
        stats.put("LA", List.of(
            "Nneka Ogwumike,14,172", "Chelsea Gray,16,224", "Candace Parker,16,211"));
        stats.put("LV", List.of(
            "A’ja Wilson,15,304", "Dearica Hamby,15,188", "Angel McCoughtry,15,220"));
        stats.put("MIN", List.of(
            "Napheesa Collier,16,262", "Crystal Dangerfield,16,254"));
        stats.put("NY", List.of(
            "Layshia Clarendon,15,18"));
        stats.put("PHX", List.of(
            "Diana Taurasi,13,236", "Brittney Griner,12,212", "Skylar Diggins-Smith,16,261",
            "Bria Hartley,13,190"));
        stats.put("SEA", List.of(
            "Breanna Stewart,16,317", "Jewell Loyd,16,223"));
        stats.put("WSH", List.of(
            "Emma Meesseman,13,158", "Ariel Atkins,15,212", "Myisha Hines-Allen,15,236"));

        assertEquals(
            Exercises.topTenScorers(stats),
            List.of(
                "Arike Ogunbowale|22.00|DAL",
                "A’ja Wilson|20.27|LV",
                "Breanna Stewart|19.81|SEA",
                "DeWanna Bonner|17.81|CONN",
                "Kelsey Mitchell|17.50|IND",
                "Betnijah Laney|16.44|ATL",
                "Napheesa Collier|16.38|MIN",
                "Skylar Diggins-Smith|16.31|PHX",
                "Crystal Dangerfield|15.88|MIN",
                "Myisha Hines-Allen|15.73|WSH"
            )
        );
    }
}